### PR TITLE
[Sprint 22 / PD-401] [Enhancement] Update Github Workflows #40

### PR DESCRIPTION
**Pull request by marfanr**
## Jira Link: 
https://ichiro-its.atlassian.net/jira/software/c/projects/PD/boards/4?selectedIssue=PD-401

## Description
update github workflow using ros2-ws-action and setup grpc on worklfows

## Type of Change

- [ ] Bugfix
- [x] Enhancement
- [ ] New feature
- [ ] Breaking change (fix or feature that would cause the existing functionality to not work as expected)

## How Has This Been Tested?

- [ ] New unit tests added.
- [ ] Manual tested.

## Checklist:

- [x] Using Branch Name Convention
    - `feature/JIRA-ID-SHORT-DESCRIPTION` if has a JIRA ticket
    - `enhancement/SHORT-DESCRIPTION` if has/has no JIRA ticket and contain enhancement
    - `hotfix/SHORT-DESCRIPTION` if the change doesn't need to be tested (urgent)
- [ ] I have commented on my code, particularly in hard-to-understand areas.
- [ ] I have made the documentation for the corresponding changes.

### CODE REVIEW

File: .github/workflows/build-ros2-ws-action.yml [hunk 1]
```diff
@@ -0,0 +1,91 @@
+name: CI
+on:
+  workflow_dispatch:
+  pull_request:
+  push:
+    branches: [master, feature/fix-github-workflows]
```
Comment (threeal) [inline .github/workflows/build-ros2-ws-action.yml]: I suggests removing the `feature/fix-github-workflows` branch from triggering the workflow. Any branch other than the main branch can be triggered by the `pull_request` event instead.
```suggestion
  push:
    branches: [master]
```

File: .github/workflows/build-ros2-ws-action.yml [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+  workflow_run:
+    workflows: [Build and Test Nightly]
+    types:  [requested, completed, in_progress]
+    branches: [feature/fix-github-workflows]
```
Comment (threeal) [inline .github/workflows/build-ros2-ws-action.yml]: I don't know why we need this, but it seems that having this branch to be triggered by the `pull_request` event is enough.
```suggestion
```

File: .github/workflows/build-ros2-ws-action.yml [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+jobs:
+  build-and-test:
+    name: Build and Test
+    runs-on: ubuntu-latest
+    steps:
+      - name: Checkout
+        uses: actions/checkout@v2.3.4
```
Comment (threeal) [inline .github/workflows/build-ros2-ws-action.yml]: Please use the latest version (see [this](https://github.com/actions/checkout/releases)):
```suggestion
        uses: actions/checkout@v4.1.1
```

File: .github/workflows/build-ros2-ws-action.yml [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+          git clone --recurse-submodules -b v1.61.0 --depth 1 --shallow-submodules https://github.com/grpc/grpc
+          mkdir -p grpc/build && cd grpc/build
+          cmake -DgRPC_INSTALL=ON -DgRPC_BUILD_TESTS=OFF ..
+          make -j 4
+          sudo make install 
+
+      - name: remove grpc and abseil clone folders
+        run: |
+          rm -rf grpc
+          rm -rf abseil-cpp
```
Comment (threeal) [inline .github/workflows/build-ros2-ws-action.yml]: We don't need to remove these directories, let's just keep it to minimize the workflow step.
```suggestion
```

File: .github/workflows/build-ros2-ws-action.yml [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+          sudo apt update -y
+          sudo apt install -y build-essential autoconf libtool pkg-config cmake
+
+      - name: Setup abseil for grpc              
+        run: |
+          git clone https://github.com/abseil/abseil-cpp.git
+          mkdir -p abseil-cpp/build && cd  abseil-cpp/build
+          cmake ..
+          make -j 4
+          sudo make install
```
Comment (threeal) [inline .github/workflows/build-ros2-ws-action.yml:69]: Use this command instead:
```suggestion
          cmake -B abseil-cpp/build
          cmake --build abseil-cpp/build
          cmake --install abseil-cpp/build
```
Also change the `Setup grpc` step.

File: .github/workflows/build-ros2-ws-action.yml [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+      - name: Checkout tachimawari interfaces
+        uses: actions/checkout@v2.3.4
+        with:
+          repository: ichiro-its/tachimawari_interfaces
+          path: tachimawari_interfaces
+
+      - name: Setup cmake        
+        run: |
+          sudo apt update -y
+          sudo apt install -y build-essential autoconf libtool pkg-config cmake
```
Comment (threeal) [inline .github/workflows/build-ros2-ws-action.yml]: It seems that we can skip the `apt update` step:
```suggestion
        run: sudo apt install -y build-essential autoconf libtool pkg-config cmake
```

File: .github/workflows/build-ros2-ws-action.yml [hunk 7]
```diff
@@ -0,0 +1,91 @@
...
+        uses: actions/checkout@v2.3.4
+        with:
+          repository: ichiro-its/tachimawari
+          path: tachimawari
+
+      - name: Checkout tachimawari interfaces
+        uses: actions/checkout@v2.3.4
+        with:
+          repository: ichiro-its/tachimawari_interfaces
+          path: tachimawari_interfaces
```
Comment (threeal) [inline .github/workflows/build-ros2-ws-action.yml]: To make it simple, maybe we can just git clone them all like what we did in the setup abseil and grpc steps. I don't know though what is the real benefit of using the `actions/checkout` compared to the `git clone` command for cloning a repository other than the current repository.

**Review by threeal — CHANGES_REQUESTED** [review]

File: .github/workflows/build-ros2-ws-action.yml (hunk 7 above)
Comment (marfanr) [reply .github/workflows/build-ros2-ws-action.yml]: Maybe `actions/checkout` will help with the ethics of wanting to clone from a private repository because it provides parameters to add an SSH key mas.

Comment (threeal) [reply .github/workflows/build-ros2-ws-action.yml]: which private repository we're going to clone in this case?

File: .github/workflows/build-ros2-ws-action.yml [hunk 8]
```diff
@@ -0,0 +1,79 @@
...
+        uses: actions/checkout@v4.1.1
+        with:
+          path: aruku
+          ref: feature/fix-using-grpc
+
+      - name: Checkout Kansei
+        uses: actions/checkout@v4.1.1
+        with:
+          repository: ichiro-its/kansei          
+          path: kansei
```
Comment (threeal) [inline .github/workflows/build-ros2-ws-action.yml]: Quick question, is it necessary to set the `path` if the value is the same as the repository name? Can you try this instead and see if it behaves the same?
```suggestion
      - name: Checkout Kansei
        uses: actions/checkout@v4.1.1
        with:
          repository: ichiro-its/kansei
```

File: .github/workflows/build-ros2-ws-action.yml (hunk 5 above)
Comment (threeal) [inline .github/workflows/build-ros2-ws-action.yml]: @marfanr you haven't implemented this for the `Setup grpc` step.

File: .github/workflows/build-ros2-ws-action.yml [hunk 9]
```diff
@@ -0,0 +1,79 @@
...
+      - name: Setup cmake        
+        run: sudo apt install -y build-essential autoconf libtool pkg-config cmake
+
+      - name: Setup abseil for grpc              
+        run: |
+          git clone https://github.com/abseil/abseil-cpp.git
+          mkdir -p abseil-cpp/build && cd  abseil-cpp/build
+          cmake ..
+          make -j 4
+          sudo make install
```
Comment (threeal) [inline .github/workflows/build-ros2-ws-action.yml]: Use this command instead:
```suggestion
          cmake -S abseil-cpp -B abseil-cpp/build 
          cmake --build abseil-cpp/build
          cmake --install abseil-cpp/build
```

File: .github/workflows/build-ros2-ws-action.yml [hunk 10]
```diff
@@ -0,0 +1,79 @@
...
+      
+      - name: Setup grpc        
+        run: |
+          git clone --recurse-submodules -b v1.61.0 --depth 1 --shallow-submodules https://github.com/grpc/grpc
+          cmake -B abseil-cpp/build
+          cmake --build abseil-cpp/build
+          cmake --install abseil-cpp/build
```
Comment (threeal) [inline .github/workflows/build-ros2-ws-action.yml]: It seems that you put this wrongly, instead of in the `Setup grpc` step, it should be in the `Setup abseil for grpc` step. But anyway there's also a bug in here, use this instead:
```suggestion
          cmake -S grpc -B grpc/build
          cmake --build grpc/build
          cmake --install grpc/build
```

**Review by threeal — CHANGES_REQUESTED** [review]
I also suggests to remove the other workflows since they are unused.

File: .github/workflows/build-ros2-ws-action.yml (hunk 7 above)
Comment (marfanr) [reply .github/workflows/build-ros2-ws-action.yml]: Oh, there isn't any in this repository, but does this also change all actions in all repositories? From those that previously used checkout to regular git clone?

Comment (marfanr) [reply .github/workflows/build-ros2-ws-action.yml]: > which private repository we're going to clone in this case?

Oh, there isn't any in this repository, but does this also change all actions in all repositories? From those that previously used checkout to regular git clone?

Comment (threeal) [reply .github/workflows/build-ros2-ws-action.yml]: It's actually up to you, and I also don't know what's best for this case.

But here's some insight:
- It seems that what makes [`actions/checkout`](https://github.com/actions/checkout) commonly used in GitHub workflows is that it handles checking out the current repository on a specific commit depending on the event (hence the name Checkout action). In this case, you may not need to use `actions/checkout` if the objective is to clone external repositories.
- In this action, we need to clone multiple repositories. In GitHub workflows, people normally use `actions/checkout` to clone repositories (including external repositories). However, if we use this, we'll have a lot of steps just for cloning the dependencies, which in my opinion looks messy.
- As an alternative, maybe we can run the step as a script and call multiple `git clone` commands in the same step. Remember that you don't need [a sledgehammer to crack a nut](https://dictionary.cambridge.org/dictionary/english/sledgehammer-to-crack-a-nut).
- There's also another alternative like [`vweevers/multi-checkout-action`](https://github.com/vweevers/multi-checkout-action) which supports checking out multiple repos in the same step and even supports private repositories.

In short, my point is not about the `git clone` command itself, but having to specify multiple steps just for cloning external dependencies seems messy to me. However, it's also fine if you prefer to keep this style.

File: .github/workflows/build-ros2-ws-action.yml [hunk 11]
```diff
@@ -0,0 +1,57 @@
...
+  pull_request:
+  push:
+    branches: [master]  
+jobs:
+  build-and-test:
+    name: Build and Test
+    runs-on: ubuntu-latest
+    steps:
+      - name: Checkout
+        run: git clone -b feature/fix-using-grpc https://github.com/ichiro-its/aruku.git
```
Comment (threeal) [inline .github/workflows/build-ros2-ws-action.yml]: Please revert this to use the Checkout action. My previous suggestion only applies to external repositories, not to the current repository. You may also need to keep the `path` input, otherwise, it will be cloned into the current working directory, which would cause the other packages to be cloned inside the Aruku package instead of having the Aruku package at the same level as the other packages.

```suggestion
      - name: Checkout
        uses: actions/checkout@v4.1.1
        with:
          path: aruku
```

File: .github/workflows/build-ros2-ws-action.yml [hunk 12]
```diff
@@ -0,0 +1,57 @@
...
+        run: git clone https://github.com/ichiro-its/aruku_interfaces.git     
+
+      - name: Checkout Keisan
+        run: git clone https://github.com/ichiro-its/keisan.git        
+
+      - name: Checkout tachimawari
+        run: git clone https://github.com/ichiro-its/tachimawari.git  
+
+      - name: Checkout tachimawari interfaces
+        run: git clone https://github.com/ichiro-its/tachimawari_interfaces.git
```
Comment (threeal) [inline .github/workflows/build-ros2-ws-action.yml]: Since you proposed using the `git clone` command, as in my previous comment, I suggest running all commands in the same step:

```suggestion
      - name: Checkout dependencies
        run: |
          git clone https://github.com/ichiro-its/kansei.git                 
          git clone https://github.com/ichiro-its/kansei_interfaces.git
          git clone https://github.com/ichiro-its/aruku_interfaces.git
          git clone https://github.com/ichiro-its/keisan.git
          git clone https://github.com/ichiro-its/tachimawari.git
          git clone https://github.com/ichiro-its/tachimawari_interfaces.git
```

**Review by threeal — CHANGES_REQUESTED** [review]

File: .github/workflows/build-ros2-ws-action.yml [hunk 13]
```diff
@@ -0,0 +1,57 @@
...
+        run: git clone https://github.com/ichiro-its/keisan.git        
+
+      - name: Checkout tachimawari
+        run: git clone https://github.com/ichiro-its/tachimawari.git  
+
+      - name: Checkout tachimawari interfaces
+        run: git clone https://github.com/ichiro-its/tachimawari_interfaces.git    
+
+      - name: Setup cmake        
+        run: sudo apt install -y build-essential autoconf libtool pkg-config cmake
```
Comment (threeal) [inline .github/workflows/build-ros2-ws-action.yml]: Can you verify if all these packages are required in this context?

It seems that most packages here are already installed in the latest Ubuntu runner (see [this](https://github.com/actions/runner-images/blob/main/images/ubuntu/Ubuntu2204-Readme.md)).

Maybe we can remove some packages from this list?

File: .github/workflows/build-ros2-ws-action.yml [hunk 14]
```diff
@@ -0,0 +1,57 @@
...
+
+      - name: Setup abseil for grpc              
+        run: |
+          git clone https://github.com/abseil/abseil-cpp.git
+          cmake -S abseil-cpp -B abseil-cpp/build 
+          cmake --build abseil-cpp/build
+          cmake --install abseil-cpp/build
```
Comment (threeal) [inline .github/workflows/build-ros2-ws-action.yml]: Maybe we need super user access for installing this?

```suggestion
          sudo cmake --install abseil-cpp/build
```

image

File: .github/workflows/build-ros2-ws-action.yml (hunk 13 above)
Comment (marfanr) [reply .github/workflows/build-ros2-ws-action.yml]: I try to remove this from this step, mas. If I encounter an error, I will put it back.

File: .github/workflows/build-ros2-ws-action.yml (hunk 7 above)
Comment (marfanr) [reply .github/workflows/build-ros2-ws-action.yml]: oh i see, okay mas I'll try to simplify everything by just using 'git clone' for the external repository.

**Review by threeal — APPROVED** [review]
The code looks good, but the CI is failing. You probably need to check that.

File: .github/workflows/dispatch-discord-pr-bot.yml [hunk 15]
```diff
@@ -0,0 +1,20 @@
...
+    steps:
+      - uses: peter-evans/repository-dispatch@v3
+        with:
+          repository: ichiro-its/${{ vars.DISCORD_PR_BOT_REPO }}
+          event-type: pull_request
+          client-payload: |
+            {
+              "repository": "${{ github.event.repository.name }}",
+              "pr_type": "${{ github.event.action || 'opened' }}"
+            }
```
Comment (threeal) [inline .github/workflows/dispatch-discord-pr-bot.yml]: Missing EoL.

File: .github/workflows/build-ros2-ws-action.yml [hunk 16]
```diff
@@ -0,0 +1,47 @@
...
+          cmake -S abseil-cpp -B abseil-cpp/build 
+          cmake --build abseil-cpp/build
+          sudo cmake --install abseil-cpp/build
+      
+      - name: Setup grpc        
+        run: |
+          git clone --recurse-submodules -b v1.61.0 --depth 1 --shallow-submodules https://github.com/grpc/grpc
+          cmake -S grpc -B grpc/build
+          cmake --build grpc/build
+          sudo cmake --install grpc/build
```
Comment (threeal) [inline .github/workflows/build-ros2-ws-action.yml:38]: Try to remove `grpc` directory after installing it and see if it can fixes failure in CI.

**Review by threeal — CHANGES_REQUESTED** [review]

**Review by FaaizHaikal — APPROVED** [review]
